File: mobile/ios/packages/town-crier-data/Sources/Repositories/APIPlanningApplicationRepository.swift
import Foundation
import TownCrierDomain

/// Fetches planning applications from the Town Crier API.
public final class APIPlanningApplicationRepository: PlanningApplicationRepository, Sendable {
  private let apiClient: URLSessionAPIClient

  public init(apiClient: URLSessionAPIClient) {
    self.apiClient = apiClient
  }

  public func fetchApplications(for zone: WatchZone) async throws
    -> [PlanningApplication] {
    let dtos: [PlanningApplicationDTO]
    do {
      dtos = try await apiClient.request(
        .get("/v1/me/watch-zones/\(zone.id.value)/applications")
      )
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
    return dtos.map { $0.toDomain() }
  }

  public func fetchApplicationsPage(
    for zone: WatchZone,
    sort: ApplicationSortOrder,
    filter: ApplicationFilter,
    cursor: String?,
    limit: Int
  ) async throws -> ApplicationPage {
    var query: [URLQueryItem] = [
      URLQueryItem(name: "sort", value: sort.rawValue),
      URLQueryItem(name: "limit", value: String(limit)),
    ]
    // Status and unread are mutually exclusive at the type level, so at most one
    // of these is ever appended — the server 400s if both arrive together. The
    // `.all` case omits both (the "All" chip / Unread off).
    switch filter {
    case .all:
      break
    case .status(let status):
      query.append(URLQueryItem(name: "status", value: status.rawValue))
    case .unread:
      query.append(URLQueryItem(name: "unread", value: "true"))
    }
    if let cursor, !cursor.isEmpty {
      query.append(URLQueryItem(name: "cursor", value: cursor))
    }

    let result: (value: [PlanningApplicationDTO], nextCursor: String?)
    do {
      result = try await apiClient.requestPaged(
        .get("/v1/me/watch-zones/\(zone.id.value)/applications", query: query)
      )
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }

    return ApplicationPage(
      applications: result.value.map { $0.toDomain() },
      nextCursor: result.nextCursor
    )
  }

  public func fetchClusters(
    for zone: WatchZone,
    viewport: MapViewport,
    zoom: Int,
    filter: ApplicationFilter
  ) async throws -> [MapCluster] {
    var query: [URLQueryItem] = [
      URLQueryItem(name: "bbox", value: Self.bboxValue(for: viewport)),
      URLQueryItem(name: "zoom", value: String(zoom)),
    ]
    // Only a status filter is meaningful for the map clusters endpoint; `.all`
    // and `.unread` send no status param (the map has no unread filter).
    if case .status(let status) = filter {
      query.append(URLQueryItem(name: "status", value: status.rawValue))
    }

    let dtos: [MapClusterDTO]
    do {
      dtos = try await apiClient.request(
        .get("/v1/me/watch-zones/\(zone.id.value)/applications/clusters", query: query)
      )
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
    return dtos.compactMap { $0.toDomain() }
  }

  /// Renders the viewport as the `bbox=west,south,east,north` value the server
  /// expects. Swift's `Double` interpolation is locale-independent (always uses
  /// `.` for the decimal separator), so this is safe regardless of device locale.
  private static func bboxValue(for viewport: MapViewport) -> String {
    "\(viewport.west),\(viewport.south),\(viewport.east),\(viewport.north)"
  }

  public func fetchApplication(by id: PlanningApplicationId) async throws -> PlanningApplication {
    let dto: PlanningApplicationDTO
    do {
      // New endpoint: /v1/applications/{authorityCode}/{**name}
      // `authorityCode` is `authority` (areaId as decimal string); `name` is the
      // PlanIt case reference. The greedy `{**name}` segment preserves slashes in
      // the reference (e.g. "22/1234/FUL"). This is a single-partition point read
      // (~1 RU) replacing the old cross-partition uid scan (tc-dzwo.1).
      dto = try await apiClient.request(
        .get("/v1/applications/\(id.authority)/\(id.name)")
      )
    } catch APIError.notFound {
      throw DomainError.applicationNotFound(id)
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
    return dto.toDomain()
  }

  public func fetchApplication(bySlug authoritySlug: String, ref: String) async throws
    -> PlanningApplication {
    let dto: PlanningApplicationDTO
    do {
      // Anonymous public read: /v1/applications/by-slug/{authoritySlug}/{**ref}.
      // The greedy `{**ref}` segment preserves slashes in the full area-prefixed
      // PlanIt name, so `ref` interpolates raw exactly like `id.name` does in the
      // by-id read above (GH #738 Slice 4).
      dto = try await apiClient.request(
        .get("/v1/applications/by-slug/\(authoritySlug)/\(ref)")
      )
    } catch APIError.notFound {
      throw DomainError.applicationNotFound(
        PlanningApplicationId(authority: authoritySlug, name: ref)
      )
    } catch let domainError as DomainError {
      throw domainError
    } catch {
      throw error.toDomainError()
    }
    return dto.toDomain()
  }
}

// MARK: - Cluster DTO

/// Wire shape for one cell of the map clusters endpoint (GH#698). A `count > 1`
/// cell has a null `applicationId` and a multi-status `statusCounts`; a
/// `count == 1` cell carries the lone application's `{authority, name}` and a
/// single-entry `statusCounts`.
///
/// An *unsplittable* multi-member cell — members coincident or closer than the
/// finest grid cell, so zoom can never separate them — additionally carries
/// `applicationIds`, the capped list of its members' `{authority, name}` ids
/// (GH#722). It is omitted/null for every splittable cell, so `members` defaults
/// to empty and existing behaviour is unchanged.
struct MapClusterDTO: Decodable, Sendable {
  let latitude: Double
  let longitude: Double
  let count: Int
  let statusCounts: [String: Int]
  let applicationId: MemberDTO?
  let applicationIds: [MemberDTO]?

  struct MemberDTO: Decodable, Sendable {
    let authority: String
    let name: String
  }

  func toDomain() -> MapCluster? {
    guard let coordinate = try? Coordinate(latitude: latitude, longitude: longitude) else {
      return nil
    }
    // Fold unknown wire states into `.unknown`, summing any collisions so the
    // per-status counts still total `count`.
    let counts = statusCounts.reduce(into: [ApplicationStatus: Int]()) { acc, pair in
      let status = ApplicationStatus(rawValue: pair.key) ?? .unknown
      acc[status, default: 0] += pair.value
    }
    let member = applicationId.map { member in
      PlanningApplicationId(authority: member.authority, name: member.name)
    }
    let members =
      applicationIds?.map { member in
        PlanningApplicationId(authority: member.authority, name: member.name)
      } ?? []
    return MapCluster(
      coordinate: coordinate, count: count, statusCounts: counts, member: member, members: members)
  }
}

// MARK: - Response DTO

struct PlanningApplicationDTO: Decodable, Sendable {
  let name: String
  let uid: String
  let areaName: String
  let areaId: Int
  let address: String
  let postcode: String?
  let description: String
  let appType: String?
  let appState: String?
  let appSize: String?
  let startDate: String?
  let decidedDate: String?
  let consultedDate: String?
  let longitude: Double?
  let latitude: Double?
  let url: String?
  let link: String?
  let lastDifferent: String
  let latestUnreadEvent: LatestUnreadEventDTO?
  /// URL-safe authority slug for the public share URL. Present on the by-id
  /// detail read and the anonymous by-slug read; absent (`omitempty`) on the
  /// list/zone endpoints, so it decodes as optional (GH #738 Slice 4).
  let authoritySlug: String?

  func toDomain() -> PlanningApplication {
    let status = ApplicationStatus(rawValue: appState ?? "") ?? .unknown
    let location = mapLocation()
    let portalUrl = url.flatMap { URL(string: $0) }
    let receivedDate = parseDate(startDate) ?? Date()
    let history = synthesizeStatusHistory(status: status, receivedDate: receivedDate)

    return PlanningApplication(
      id: PlanningApplicationId(authority: String(areaId), name: name),
      reference: ApplicationReference(name),
      authority: LocalAuthority(code: String(areaId), name: areaName, slug: authoritySlug),
      status: status,
      receivedDate: receivedDate,
      description: description,
      address: address,
      location: location,
      portalUrl: portalUrl,
      statusHistory: history,
      latestUnreadEvent: latestUnreadEvent?.toDomain()
    )
  }

  private func mapLocation() -> Coordinate? {
    guard let lat = latitude, let lon = longitude else { return nil }
    return try? Coordinate(latitude: lat, longitude: lon)
  }

  private func synthesizeStatusHistory(
    status: ApplicationStatus,
    receivedDate: Date
  ) -> [StatusEvent] {
    var history: [StatusEvent] = []

    if startDate != nil {
      history.append(StatusEvent(status: .undecided, date: receivedDate))
    }

    if let decidedDateString = decidedDate, let decidedDate = parseDate(decidedDateString) {
      let decidedStatus: ApplicationStatus
      switch status {
      case .permitted, .conditions, .rejected, .withdrawn, .appealed:
        decidedStatus = status
      default:
        decidedStatus = .undecided
      }
      if decidedStatus != .undecided {
        history.append(StatusEvent(status: decidedStatus, date: decidedDate))
      }
    }

    return history
  }

  private static let dateFormatter: DateFormatter = {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd"
    formatter.locale = Locale(identifier: "en_US_POSIX")
    formatter.timeZone = TimeZone(identifier: "UTC")
    return formatter
  }()

  private func parseDate(_ dateString: String?) -> Date? {
    guard let dateString else { return nil }
    return Self.dateFormatter.date(from: dateString)
  }
}

/// Wire shape for the per-row `latestUnreadEvent` descriptor returned by the
/// per-zone applications endpoint (tc-1nsa.3). Older API builds may omit the
/// field entirely; absence and explicit `null` both decode to `nil` on the
/// domain entity. Spec: `docs/specs/notifications-unread-watermark.md`.
struct LatestUnreadEventDTO: Decodable, Sendable {
  let type: String
  let decision: String?
  let createdAt: String

  func toDomain() -> LatestUnreadEvent? {
    // Server may emit either fractional or integer-second precision so
    // try fractional first and fall back to plain ISO-8601 — both are
    // legal ISO-8601 date-time outputs the API can produce.
    let withFractional = ISO8601DateFormatter()
    withFractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let parsed = withFractional.date(from: createdAt) {
      return LatestUnreadEvent(type: type, decision: decision, createdAt: parsed)
    }
    let plain = ISO8601DateFormatter()
    plain.formatOptions = [.withInternetDateTime]
    guard let parsed = plain.date(from: createdAt) else {
      return nil
    }
    return LatestUnreadEvent(type: type, decision: decision, createdAt: parsed)
  }
}

File: mobile/android/data/src/main/kotlin/uk/towncrierapp/data/applications/ApiPlanningApplicationRepository.kt
package uk.towncrierapp.data.applications

import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import uk.towncrierapp.data.api.ApiClient
import uk.towncrierapp.data.api.ApiEndpoint
import uk.towncrierapp.data.api.DotNetTimeParser
import uk.towncrierapp.domain.applications.ApplicationFilter
import uk.towncrierapp.domain.applications.ApplicationPage
import uk.towncrierapp.domain.applications.ApplicationSortOrder
import uk.towncrierapp.domain.applications.ApplicationStatus
import uk.towncrierapp.domain.applications.LatestUnreadEvent
import uk.towncrierapp.domain.applications.LocalAuthority
import uk.towncrierapp.domain.applications.PlanningApplication
import uk.towncrierapp.domain.applications.PlanningApplicationId
import uk.towncrierapp.domain.applications.PlanningApplicationRepository
import uk.towncrierapp.domain.applications.StatusEvent
import uk.towncrierapp.domain.applications.isDecided
import uk.towncrierapp.domain.applications.wireValue
import uk.towncrierapp.domain.watchzones.Coordinate
import uk.towncrierapp.domain.watchzones.WatchZoneId
import java.time.LocalDate

/**
 * `PlanningApplicationRepository` over the Town Crier API. Always sends
 * `sort` (epic #770: Android never exercises the legacy param-less path) and
 * `limit=`[PAGE_SIZE]; `status`/`unread` are mutually exclusive by
 * [ApplicationFilter]'s shape, not a runtime check. Continuation is the
 * opaque `X-Next-Cursor` response header via [ApiClient.requestPaged]. Port
 * of iOS `APIApplicationRepository` (GH#775).
 */
public class ApiPlanningApplicationRepository(
    private val apiClient: ApiClient,
) : PlanningApplicationRepository {
    override suspend fun applications(
        zoneId: WatchZoneId,
        sort: ApplicationSortOrder,
        filter: ApplicationFilter,
        cursor: String?,
    ): ApplicationPage {
        val query =
            buildList {
                add("sort" to sort.wireValue)
                add("limit" to PAGE_SIZE.toString())
                when (filter) {
                    ApplicationFilter.All -> Unit
                    is ApplicationFilter.Status -> add("status" to filter.status.wireValue)
                    ApplicationFilter.Unread -> add("unread" to "true")
                }
                cursor?.let { add("cursor" to it) }
            }
        val result =
            apiClient.requestPaged(
                ApiEndpoint.get("/v1/me/watch-zones/${zoneId.value}/applications", query = query),
                ListSerializer(PlanningApplicationRowDto.serializer()),
            )
        return ApplicationPage(result.value.map { it.toDomain() }, result.nextCursor)
    }

    override suspend fun detail(
        authority: String,
        name: String,
    ): PlanningApplication =
        apiClient
            .request(ApiEndpoint.get("/v1/applications/$authority/$name"), ApplicationDetailDto.serializer())
            .toDomain()

    public companion object {
        public const val PAGE_SIZE: Int = 150
    }
}

// MARK: - DTOs

@Serializable
internal data class LatestUnreadEventDto(
    val type: String,
    val decision: String? = null,
    val createdAt: String,
)

@Serializable
internal data class PlanningApplicationRowDto(
    val name: String,
    val uid: String,
    val areaName: String,
    val areaId: Int,
    val address: String,
    val postcode: String? = null,
    val description: String,
    val appType: String? = null,
    val appState: String? = null,
    val appSize: String? = null,
    val startDate: String? = null,
    val decidedDate: String? = null,
    val longitude: Double? = null,
    val latitude: Double? = null,
    val url: String? = null,
    val link: String? = null,
    val lastDifferent: String,
    val latestUnreadEvent: LatestUnreadEventDto? = null,
)

@Serializable
internal data class ApplicationDetailDto(
    val name: String,
    val uid: String,
    val areaName: String,
    val areaId: Int,
    val address: String,
    val postcode: String? = null,
    val description: String,
    val appType: String? = null,
    val appState: String? = null,
    val appSize: String? = null,
    val startDate: String? = null,
    val decidedDate: String? = null,
    val longitude: Double? = null,
    val latitude: Double? = null,
    val url: String? = null,
    val link: String? = null,
    val lastDifferent: String,
    val latestUnreadEvent: LatestUnreadEventDto? = null,
    val authoritySlug: String? = null,
    val areaType: String? = null,
)

internal fun PlanningApplicationRowDto.toDomain(): PlanningApplication {
    val status = resolveStatus(appState)
    return PlanningApplication(
        id = PlanningApplicationId(authority = areaId.toString(), name = name),
        reference = name,
        authority = LocalAuthority(code = areaId.toString(), name = areaName),
        status = status,
        receivedDate = resolveReceivedDate(startDate, lastDifferent),
        description = description,
        address = address,
        location = resolveLocation(latitude, longitude),
        // "link" (the external council-portal URL) takes priority over "url"
        // (PlanIt's own page) — the "View on Council Portal" action wants the
        // former (GH#775).
        portalUrl = link ?: url,
        statusHistory = synthesizeStatusHistory(status, startDate, decidedDate),
        latestUnreadEvent = latestUnreadEvent?.toDomain(),
    )
}

internal fun ApplicationDetailDto.toDomain(): PlanningApplication {
    val status = resolveStatus(appState)
    return PlanningApplication(
        id = PlanningApplicationId(authority = areaId.toString(), name = name),
        reference = name,
        authority = LocalAuthority(code = areaId.toString(), name = areaName, areaType = areaType, slug = authoritySlug),
        status = status,
        receivedDate = resolveReceivedDate(startDate, lastDifferent),
        description = description,
        address = address,
        location = resolveLocation(latitude, longitude),
        portalUrl = link ?: url,
        statusHistory = synthesizeStatusHistory(status, startDate, decidedDate),
        latestUnreadEvent = latestUnreadEvent?.toDomain(),
    )
}

// A missing/unparseable createdAt drops just the unread indicator, not the
// whole row — a stale or malformed event server-side shouldn't hide an
// otherwise-good application (GH#775).
internal fun LatestUnreadEventDto.toDomain(): LatestUnreadEvent? =
    DotNetTimeParser.parse(createdAt)?.let { LatestUnreadEvent(type = type, decision = decision, createdAt = it) }

// An absent appState is wire-legitimate (optional field) — treated the same
// as any other unrecognised raw value: Unknown, never a crash.
private fun resolveStatus(appState: String?): ApplicationStatus = appState?.let(ApplicationStatus::fromWireValue) ?: ApplicationStatus.Unknown("")

// startDate is the primary source; lastDifferent (always present) is the
// fallback for the rare row missing it — receivedDate is non-nullable
// domain-side and must resolve to SOMETHING (GH#775).
private fun resolveReceivedDate(
    startDate: String?,
    lastDifferent: String,
): LocalDate =
    startDate?.let(DotNetTimeParser::parseDate)
        ?: DotNetTimeParser.parse(lastDifferent)?.toLocalDate()
        ?: LocalDate.EPOCH

private fun resolveLocation(
    latitude: Double?,
    longitude: Double?,
): Coordinate? = if (latitude != null && longitude != null) Coordinate(latitude, longitude) else null

// Max two points: startDate -> Undecided, decidedDate -> the actual decided
// status, but ONLY when that status really is decided — a decidedDate on a
// still-Unresolved/Referred/etc. row is folded away (GH#775).
private fun synthesizeStatusHistory(
    status: ApplicationStatus,
    startDate: String?,
    decidedDate: String?,
): List<StatusEvent> =
    buildList {
        startDate?.let(DotNetTimeParser::parseDate)?.let { add(StatusEvent(ApplicationStatus.Undecided, it)) }
        if (status.isDecided) {
            decidedDate?.let(DotNetTimeParser::parseDate)?.let { add(StatusEvent(status, it)) }
        }
    }
